tests/mongo-tools/dump_restore_test.go

```diff
@@ -15,63 +15,112 @@
-	"go.mongodb.org/mongo-driver/bson"
-
-	"github.com/FerretDB/dance/tests/common"
 )
 
 func TestDumpRestore(t *testing.T) {
-	ctx, db := common.Setup(t)
+	t.Parallel()
 
-	// TODO restart existing dump from https://github.com/mcampo2/mongodb-sample-databases instead
-	_, err := db.Collection("test").InsertOne(ctx, bson.D{{"foo", "bar"}})
-	require.NoError(t, err)
+	containerSourceRoot := "/dumps/mongodb-sample-databases/dump"
+	localTestsRoot := filepath.Join("..", "..", "dumps", "tests")
+	containerTestsRoot := "/dumps/tests"
 
-	localRoot := filepath.Join("..", "..", "dumps")
-	containerRoot := "/dumps/"
+	//nolint:vet // for readability
+	type testCase struct {
+		name0          string
+		documentsCount map[string]int // collection name -> document count
+	}
 
-	expectedState := getDatabaseState(t, ctx, db)
+	for _, tc := range []testCase{{
+		name0: "sample_geospatial",
+		documentsCount: map[string]int{
+			"shipwrecks": 11095,
+		},
+	}, {
+		name0: "sample_analytics",
+		documentsCount: map[string]int{
+			"accounts":     1746,
+			"customers":    500,
+			"transactions": 1746,
+		},
+	}} {
+		tc := tc
+		t.Run(tc.name0, func(t *testing.T) {
+			t.Parallel()
 
-	// pre-create directories to avoid permission issues
-	err = os.Chmod(localRoot, 0o777)
-	require.NoError(t, err)
-	err = os.RemoveAll(filepath.Join(localRoot, db.Name()))
-	require.NoError(t, err)
-	err = os.Mkdir(filepath.Join(localRoot, db.Name()), 0o777) // 0o777 is typically downgraded to 0o755 by umask
-	require.NoError(t, err)
-	err = os.Chmod(filepath.Join(localRoot, db.Name()), 0o777) // fix after umask
-	require.NoError(t, err)
+			name1 := tc.name0 + "_dump1"
+			name2 := tc.name0 + "_dump2"
 
-	// dump a database
-	err = runDockerComposeCommand(
-		"mongodump",
-		"--out", containerRoot,
-		"--db", db.Name(),
-		"--verbose",
-		"mongodb://host.docker.internal:27017/",
-	)
-	require.NoError(t, err)
+			// pre-create directory to avoid permission issues
+			recreateDir(t, filepath.Join(localTestsRoot, name1))
+
+			ctx, client := setup(t)
+
+			db1 := client.Database(name1)
+			t.Cleanup(func() { require.NoError(t, db1.Drop(ctx)) })
+
+			db2 := client.Database(name2)
+			t.Cleanup(func() { require.NoError(t, db2.Drop(ctx)) })
+
+			// source dump -> db1
+			mongorestore(t, tc.name0, containerSourceRoot, name1)
+			actualCount := getDocumentsCount(t, ctx, db1)
+			assert.Equal(t, tc.documentsCount, actualCount)
+
+			// db1 -> test dump
+			mongodump(t, name1, containerTestsRoot)
 
-	// cleanup database
-	ctx, db = common.Setup(t)
+			// test dump -> db2
+			mongorestore(t, name1, containerTestsRoot, name2)
+			actualCount = getDocumentsCount(t, ctx, db2)
+			assert.Equal(t, tc.documentsCount, actualCount)
 
-	// restore a database based on created dump
-	err = runDockerComposeCommand(
+			compareDatabases(t, ctx, db1, db2)
+
+			// we can't compare bson dump files because `mongodump` queries documents in natural order,
+			// but we don't support it
```

Comment:
I think I might still don't understand why we are unable to generate same dumps. The "bootstrapping" approach was able to generate 2 same dumps as the order on the same database is preserved. Why we removed it?

Reply:
Because it doesn't work for the second test with a slightly bigger database. The fact that the file was exactly the same for a smaller database was a happy accident